### PR TITLE
Databases pages

### PR DESCRIPTION
**Pull request by maxlang**
- changes the way the REST api is specified to be consistent with
  the rest of the application
- adds proxyquire for requiring modules with their dependencies
  substituted with custom mocks
- lots of boilerplate that will hopefully be removed in future versions

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/6755)

### CODE REVIEW

File: ui/next/app/redux/databaseInfo.ts [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+      let { payload: id } = action as PayloadAction<string>;
+      state = _.clone(state);
+      state[id] = _.clone(state[id]) || new DatabaseDetailsState();
+      state[id].inFlight = true;
+      return state;
+    case DATABASE_DETAILS_RECEIVE:
+      // The results of a request have been received.
+      let { payload } = action as PayloadAction<WithID<DatabaseDetailsResponseMessage>>;
+      state = _.clone(state);
+      state[payload.id] = _.clone(state[payload.id]) || new DatabaseDetailsState();
```
Comment (mrtracy) [inline ui/next/app/redux/databaseInfo.ts]: This is not how this should be done - you should have a per-item reducer.  Look at the collection in metrics.ts.

It should be something like:
```typescript
case DATABASE_DETAILS_RECEIVE:
case DATABASE_DETAILS_ERROR:
case DATABASE_DETAILS_INVALIDATE:
  let { payload } = action as PayloadAction<WithID<any>>;
  let state = _.clone(state);
  state[id] = singleDetailReducer(state[id], action);
  return state;
```

**Comment by mrtracy** [comment]
I agree with your assertion that this is a pretty exceptional amount of copy/paste.  I think its okay to check in for the purpose of getting ui/next installed, but this would certainly be worth revisiting in the near term.

File: ui/next/app/redux/databaseInfo.ts (hunk 1 above)
Comment (maxlang) [inline ui/next/app/redux/databaseInfo.ts]: What's the benefit of this way over the other? Since essentially the same thing is taking place.

**Comment by maxlang** [comment]
Even though there was a lot of copy-paste, each page did have a few quirks to work out. I should be able to simplify on the next pass though.

There were a number of edge cases that made not-copying tricky - I actually merged and umerged some reducers reducers and was automatically generating table columns, but this would interfere alternately with typescript typings, the redux action constants, or typescript enums making it less straightforward than plain JS.

**Comment by mrtracy** [comment]
Let's get this checked in, but there are considerable opportunities for useful abstraction here.  LGTM.

<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Even though there was a lot of copy-paste, each page did have a few quirks to work out. I should be able to simplify on the next pass though.
> 
> 
> 
> There were a number of edge cases that made not-copying tricky - I actually merged and umerged some reducers and was automatically generating table columns, but this would interfere alternately with typescript typings, the redux action constants, or typescript enums making it less straightforward than plain JS.
</details>


---

Reviewed 1 of 13 files at r2.
Review status: 1 of 15 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[ui/next/app/containers/databases/databaseDetails.tsx, line 174 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6755#-KIVYVT0pi7TDN6G5NOr:-KIVYVT0pi7TDN6G5NOs:874790118) ([raw file](https://github.com/cockroachdb/cockroach/blob/de79b113bd07227678c3f7c8b74ba2d0f95dc917/ui/next/app/containers/databases/databaseDetails.tsx#L174)):*
> ```XML
>   /**
>    * tableColumns is a selector which computes the input Columns to the table table,
>    * based the tableColumnDescriptors and the current sorted table data
> ```

s/based/based on/

---

*[ui/next/app/containers/databases/databaseDetails.tsx, line 306 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6755#-KIVWnRI_ooy-kUFXV5y:-KIVWnRI_ooy-kUFXV5z:480988498) ([raw file](https://github.com/cockroachdb/cockroach/blob/de79b113bd07227678c3f7c8b74ba2d0f95dc917/ui/next/app/containers/databases/databaseDetails.tsx#L306)):*
> ```XML
>       refreshDatabaseDetails: function () {
>         // create a new function that applies the appropriate database name to the refreshDatabaseDetails function
>         dispatch((() => refreshDatabaseDetails(ownProps.params.database_name)).apply(undefined, arguments));
> ```

I find this to be extremely confusing, and I don't know why it is necessary.  Why not just use the simpler `connect` syntax used in all of our other components?  `componentDidMount` has access to the database name, it can call `this.props.refreshDatabaseDetails(this.props.params.database_name)`.

---

*[ui/next/app/redux/databaseInfo.spec.ts, line 40 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6755#-KIV_-Lkzea4HuJZiQhV:-KIV_-Lkzea4HuJZiQhW:-317042304) ([raw file](https://github.com/cockroachdb/cockroach/blob/de79b113bd07227678c3f7c8b74ba2d0f95dc917/ui/next/app/redux/databaseInfo.spec.ts#L40)):*
> ```XML
>   });
> 
>   describe("databse details actions", function () {
> ```

database

---

*[ui/next/app/redux/databaseInfo.spec.ts, line 391 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6755#-KIVa-kwgMxxt-BuJ2TO:-KIVa-kwgMxxt-BuJ2TP:-279623377) ([raw file](https://github.com/cockroachdb/cockroach/blob/de79b113bd07227678c3f7c8b74ba2d0f95dc917/ui/next/app/redux/databaseInfo.spec.ts#L391)):*
> ```XML
>             getDatabaseList,
>           },
>         });
> ```

I don't like the use of proxyquire, it's a fairly complicated mock for an object we have complete control over.  This is acceptable to get this checked in, but I would have added an additional, optional parameter to the action creators.

Something like `refreshDatabases(getDatabaseList = API.getDatabaseList)`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6755#-:-KIVgIMaYahL1lQYRTUl:-1520346688)*

**Comment by maxlang** [comment]
Sounds good.

<details><summary><i>Previously, mrtracy wrote…</i></summary>
> Let's get this checked in, but there are considerable opportunities for useful abstraction here.  LGTM.
</details>


---

Review status: 1 of 15 files reviewed at latest revision, 5 unresolved discussions.

---

*[ui/next/app/containers/databases/databaseDetails.tsx, line 174 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6755#-KIVYVT0pi7TDN6G5NOr:-KIY1i4wrS5V2AizEM9G:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/de79b113bd07227678c3f7c8b74ba2d0f95dc917/ui/next/app/containers/databases/databaseDetails.tsx#L174)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> s/based/based on/
</details>
Done.

---

*[ui/next/app/containers/databases/databaseDetails.tsx, line 306 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6755#-KIVWnRI_ooy-kUFXV5y:-KIY3211-BCO15f5JNjH:-1530787867) ([raw file](https://github.com/cockroachdb/cockroach/blob/de79b113bd07227678c3f7c8b74ba2d0f95dc917/ui/next/app/containers/databases/databaseDetails.tsx#L306)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> I find this to be extremely confusing, and I don't know why it is necessary.  Why not just use the simpler `connect` syntax used in all of our other components?  `componentDidMount` has access to the database name, it can call `this.props.refreshDatabaseDetails(this.props.params.database_name)`.
</details>
Ah yeah, of course.

---

*[ui/next/app/redux/databaseInfo.spec.ts, line 40 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6755#-KIV_-Lkzea4HuJZiQhV:-KIYEnITAvrCromcXYtI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/de79b113bd07227678c3f7c8b74ba2d0f95dc917/ui/next/app/redux/databaseInfo.spec.ts#L40)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> database
</details>
Done.

---

*[ui/next/app/redux/databaseInfo.spec.ts, line 391 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6755#-KIVa-kwgMxxt-BuJ2TO:-KIYF6d0CC3Vl4Mq76tE:-225507499) ([raw file](https://github.com/cockroachdb/cockroach/blob/de79b113bd07227678c3f7c8b74ba2d0f95dc917/ui/next/app/redux/databaseInfo.spec.ts#L391)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> I don't like the use of proxyquire, it's a fairly complicated mock for an object we have complete control over.  This is acceptable to get this checked in, but I would have added an additional, optional parameter to the action creators.
> 
> Something like `refreshDatabases(getDatabaseList = API.getDatabaseList)`.
</details>
We can discuss this in person. I'd probably rather the extra complexity live in the tests, but for a simpler case like this I could see it going both ways.

---

*[ui/next/app/redux/databaseInfo.ts, line 126 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6755#-KI-Lsap1T32UA1Hev4P-r1-126:-KIYFlqL-yfQbCGDG0yI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/72410a488bd1067bc5ec9cdc9b0a2101f502250b/ui/next/app/redux/databaseInfo.ts#L126)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> What's the benefit of this way over the other? Since essentially the same thing is taking place.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6755#-:-KIY1b_5JqqTpxZCMhRz:1689193178)*

**Comment by maxlang** [comment]
Circle CI isn't finding it

**Pull request closed without merging**